Improve documentation

SergioGasquez opened the pull request:
We can probably use the same documentation in the readme, so I will wait until I get some feedback, and then I will copy it to the readme.

bugadani commented on src/lib.rs:
```diff
@@ -1,3 +1,66 @@
+//! - `critical-section` enables critical sections.
+//! - `colors` enable colored output.
+//! - There is one feature for each [`defmt` supported encoding]:
+//!    - `defmt`: Uses [rzCOBS] encoding
+//!    - `defmt-raw`: Raw data, that no encoding.
```
```suggestion
//!    - `defmt`: Uses [rzCOBS] encoding and adds framing. This is intended to be used with `espflash`
//!    - `defmt-raw`: Raw defmt output without additional framing.
```

bugadani replied:
The point of these isn't the encoding, but the additional framing used for `defmt` & `espflash`

bugadani commented on src/lib.rs:
```diff
@@ -1,3 +1,66 @@
+//!
+//! | `esp-println` | `esp-backtrace`      |
+//! | ------------- | -------------------- |
+//! | `uart`        | `print- uart`        |
+//! | `jtag-serial` | `print- jtag-serial` |
+//! | `rtt`         | `print- rtt`         |
```
```suggestion
//! | `uart`        | `print-uart`        |
//! | `jtag_serial` | `print-jtag-serial` |
//! | `rtt`         | `print-rtt`         |
```

There's a slight inconsistency here - but do we really want these features on esp-backtrace?

bjoernQ commented:
Looks good!

> We can probably use the same documentation in the readme, so I will wait until I get some feedback, and then I will copy it to the readme.

Maybe we can avoid duplication with something like `#![doc = include_str!("README.md")]` and only have it as the README.md?

bugadani commented on src/lib.rs:
```diff
@@ -1,3 +1,69 @@
+//!   `jtag-serial` and `rtt` `no-op`.
+//!     - Only one of these features can be enabled at a time.
+//! - `critical-section` enables critical sections.
+//! - `colors` enable colored output.
```
`colors` only affects `log` I believe. We might want to add this as a note here, but we could also add a warning to `build.rs` to make this obvious. Thoughts?

SergioGasquez replied:
From what I've seen, you are right. So I added the `build.rs` check and updated the documentation

bugadani commented on src/lib.rs:
```diff
@@ -1,3 +1,69 @@
+//!     - Only one of these features can be enabled at a time.
+//! - `critical-section` enables critical sections.
+//! - `colors` enable colored output.
+//! - There is one feature for each [`defmt` supported encoding]:
+//!    - `defmt`: Uses [rzCOBS] encoding and adds framing. This is intended to
+//!      be used with [`espflash`] (See `-L/--log-format` argument of `flash` or
+//!      `monitor` subcommands of `espflash` and `cargo-espflash`)
+//!    - `defmt-raw`: Raw defmt output without additional framing.
```
```suggestion
//! - There are two ways to use this library with [`defmt`]:
//!    - `defmt`: This is intended to be used with [`espflash`]
//!      See `-L/--log-format` argument of `flash` or `monitor`
//!      subcommands of `espflash` and `cargo-espflash`). Uses
//!      [rzCOBS] encoding and adds framing.
//!    - `defmt-raw`: Raw defmt output without additional framing.
//!      ⚠️ Be careful when using this feature: you must only write
//!      output using `defmt` macros, or you may irrecoverably corrupt
//!      the output stream! This means that even the bootloader's output
//!      must be disabled.
```

bugadani replied:
Sorry I was a bit lazy with the previous comment here. This change would make the ``[`defmt` supported encoding]`` link unused.

bugadani commented on src/lib.rs:
```diff
@@ -1,3 +1,69 @@
+//! [`defmt` supported encoding]: https://defmt.ferrous-systems.com/encoding.html?highlight=rzcobs#encoding
+//! [rzCOBS]: https://github.com/Dirbaio/rzcobs
+//! [`espflash`]: https://github.com/esp-rs/espflash
+//! [rtt]: https://crates.io/crates/espflash
```
Does `espflash` support rtt, or is this a copy-paste bug?

SergioGasquez commented:
> Maybe we can avoid duplication with something like #![doc = include_str!("README.md")] and only have it as the README.md?

TIL! 

> Does espflash support rtt, or is this a copy-paste bug?

Copy-paste error :sweat:

Moved the documentation the README and addressed some feedback

SergioGasquez commented on src/lib.rs:
```diff
@@ -1,3 +1,69 @@
+//!    - Only one of these features can be enabled at a time.
+//! - There is one feature for each supported communication method: `uart`,
+//!   `jtag-serial` and `rtt` `no-op`.
+//!     - Only one of these features can be enabled at a time.
```
Right, but this changes from chip to chip, devkit to devkit, any idea on how to document this. Any idea how to improve it?

bjoernQ reviewed (approved):
LGTM

MabezDev commented on README.md:
```diff
@@ -1,31 +1,59 @@
+- There are two ways to use this library with [`defmt`]:
+   - `defmt`: This is intended to be used with [`espflash`], see `-L/--log-format` argument of `flash` or `monitor` subcommands of `espflash` and `cargo-espflash`
+     Uses [rzCOBS] encoding and adds framing.
+   - `defmt-raw`: Raw defmt output without additional framing. ⚠️ Be careful when using this feature: you must only write
```
I guess we should also mention probe-rs here? Based on the discussion in https://github.com/esp-rs/esp-println/issues/65, I guess we say if you want defmt over rtt to 

- Use defmt-rtt with the defmt macros
- Use panic-probe instead of esp-backtrace

Maybe we should remove rtt support in esp-println :thinking:

bjoernQ replied:
> Maybe we should remove rtt support in esp-println

mhhhh - IIRC I only added it because there were some problems with https://crates.io/crates/rtt-target and it was unmaintained back then. But that changed 🤔

bjoernQ replied:
wouldn't mind see it go away

jessebraham replied:
Does `panic-probe` actually work with our chips? I seem to recall trying and was unable to get it building. Perhaps I'm just misremembering, though.

jessebraham replied:
Just tried again:

```text
error: `panic-probe` only supports Cortex-M targets (thumbvN-none-eabi[hf])
  --> /Users/jesse/.cargo/registry/src/index.crates.io-6f17d22bba15001f/panic-probe-0.3.1/src/lib.rs:24:1
   |
24 | compile_error!("`panic-probe` only supports Cortex-M targets (thumbvN-none-eabi[hf])");
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: could not compile `panic-probe` (lib) due to previous error
warning: build failed, waiting for other jobs to finish...
```

Haven't looked into the source yet, so not sure how much work is required to support different archs.

SergioGasquez replied:
I just mentioned probe-rs and this section, is there anything else we should address in this PR? I guess the discussion of dropping rtt support should take place in https://github.com/esp-rs/esp-println/issues/65

MabezDev replied:
Ugh, thats annoying, I looked at the code, only the "induce hard fault bit" is cortex-m specific. That could be put behind a feature. With that said, it seems defmt is getting less and less attention and they're offloading and deprecating many things. We can continue this in #65 like @SergioGasquez suggested.

jessebraham commented on README.md:
```diff
@@ -44,29 +72,50 @@ In this case the following environment variables are used:
 import statements. This ensures that the global logger will not be removed by the compiler.
 
-## License
+# License
```
Just curious about this change, it was consistent with all (most?) of the other repositories as it was, which we're not deviating from. Not a huge deal but, though.

SergioGasquez replied:
Changed it so it does look like it's a subsection of the previous h1 header. Happy to revert it if you think we should keep it as other repos.

jessebraham replied:
Yeah that's fine, as I said not a big deal. Just without context seemed like kind of a strange change 😁

jessebraham commented on README.md:
```diff
@@ -75,7 +124,7 @@ Licensed under either of:
 at your option.
 
-### Contribution
+# Contribution
```
See above.

bugadani commented on README.md:
```diff
@@ -1,31 +1,59 @@
+   - Only effective when using the `log` feature.
+- `critical-section` enables critical sections.
+- There are two ways to use this library with [`defmt`]:
+   - `defmt`: This is intended to be used with [`espflash`], see `-L/--log-format` argument of `flash` or `monitor` subcommands of `espflash` and `cargo-espflash`
```
Sorry I snuck in a rename, the feature is now called `defmt-espflash`

bugadani replied:
```suggestion
   - `defmt-espflash`: This is intended to be used with [`espflash`], see `-L/--log-format` argument of `flash` or `monitor` subcommands of `espflash` and `cargo-espflash`
```

bugadani commented on README.md:
```diff
@@ -44,29 +72,50 @@ In this case the following environment variables are used:
 
 Using the `defmt-espflash` feature, esp-println will install a defmt global logger. The logger will
 output to the same data stream as `println!()`, and adds framing bytes so it can be used even with
 other, non-defmt output. Using the `defmt` feature automatically uses the Rzcobs encoding and does
```
I missed a rename here
```suggestion
other, non-defmt output. Using the `defmt-espflash` feature automatically uses the Rzcobs encoding and does
```

jessebraham commented on README.md:
```diff
@@ -1,31 +1,63 @@
+
+## Default Features
+
+By default, we use the `uart`, `critial-section` and `colors` features.
```
If the `colors` feature is enabled by default, and the `log` feature is required and _not_ a default feature, is this not problematic?

bjoernQ replied:
no - without the log feature colors is no-op

jessebraham replied:
The build script will panic if `colors` is enabled and `log` is not in its current state:
https://github.com/esp-rs/esp-println/pull/68/files#diff-d0d98998092552a1d3259338c2c71e118a5b8343dd4703c0c7f552ada7f9cb42R47-R50

bjoernQ replied:
ah lol - yeah introduced in this PR

probably we shouldn't have that checked or at least lower it to a warning

jessebraham commented on build.rs:
```diff
@@ -43,4 +43,9 @@ fn main() {
+
+    // Ensure that, if the `colors` is used with `log`.`
+    if cfg!(feature = "colors") && !cfg!(feature = "log") {
+        panic!("The `colors` feature is only effective when using the `log` feature");
```
```suggestion
        println!("cargo:warning=The `colors` feature is only effective when using the `log` feature");
```

We should change this to print a warning rather than panicking.
